### DISCUSSION TITLE
How do I get the bindings to work in Windows?

### DISCUSSION THREAD

**Discussion by PlatinumLucario** [gircore/gir.core#821] [Q&A]
Hi, I've been trying to get the 0.3.0 bindings to work in Windows, and I have two questions regarding getting the bindings to work with libgtk-4-1.dll and libadwaita-1.dll and other libraries in Windows.

I'm absolutely sure I've read each step regarding installing the packages via MSYS2 MinGW and putting ``C:\msys64\mingw64\bin`` into the Environment Variable at the front of the system PATH, as seen here:
![image](https://user-images.githubusercontent.com/1906189/222886239-7e1cdb78-6d71-43dc-adf8-f5af3ea76736.png)

And the libraries are installed, as seen here:
![image](https://user-images.githubusercontent.com/1906189/222886802-90f0f4f6-4683-4f0c-aa69-04b9df59ffcb.png)

But despite doing this, when running the compiled GTK4 project I'm working on in Visual Studio 2022, it's still unable to find the libgtk-4-1 library dependency that the binding relies on.
![image](https://user-images.githubusercontent.com/1906189/222886163-2a1b18c7-130a-4814-a2cf-064ea1ed072d.png)

Surely I must be missing a step in the procedure, because normally the compiled C libraries are meant to be seen by the bindings, if they're in the correct directory.

Also, my other question is this: Is there a way to place the GTK4 C libraries into the specific project directory and make the bindings read directly from the specific project directory when running in Windows, like as if they're a .dll reference?

**Comment by badcel** [comment]
I don't have a windows device to test this. For me your setup looks ok.

You should be able to copy the c binaries into your `bin/debug` folder (next to your own exe) to ensure that it is not a path related issue.

**Reply by PlatinumLucario** [reply]
I also forgot to mention, I've already attempted placing them next to the compiled executable, and even though it takes a bit longer to reach the exception, it still occurs, as attempted again here: 
![image](https://user-images.githubusercontent.com/1906189/222896407-ec35e819-0ec8-4f9d-ae3d-550166a68c1c.png)

**Reply by badcel** [reply]
Are the libraries in your screenshot all libraries you copied over?

The error mentions that it could be that not GTK is not found but one of it's dependencies. Can you provide a full list of the c libraries you copied over?

**Reply by PlatinumLucario** [reply]
I can confirm that these are all the libraries I copied over to into the executable directory, are there any dependencies missing in this list?
```
libadwaita-1-0.dll
libcairo-2.dll
libcairo-gobject-2.dll
libcairo-script-interpreter-2.dll
libgio-2.0-0.dll
libglib-2.0-0.dll
libgobject-2.0-0.dll
libgraphene-1.0-0.dll
libgraphite2.dll
libgsl-27.dll
libgslcblas-0.dll
libgstadaptivedemux-1.0-0.dll
libgstallocators-1.0-0.dll
libgstapp-1.0-0.dll
libgstaudio-1.0-0.dll
libgstbadaudio-1.0-0.dll
libgstbase-1.0-0.dll
libgstbasecamerabinsrc-1.0-0.dll
libgstcheck-1.0-0.dll
libgstcodecparsers-1.0-0.dll
libgstcodecs-1.0-0.dll
libgstcontroller-1.0-0.dll
libgstcuda-1.0-0.dll
libgstd3d11-1.0-0.dll
libgstfft-1.0-0.dll
libgstgl-1.0-0.dll
libgstinsertbin-1.0-0.dll
libgstisoff-1.0-0.dll
libgstmpegts-1.0-0.dll
libgstnet-1.0-0.dll
libgstpbutils-1.0-0.dll
libgstphotography-1.0-0.dll
libgstplay-1.0-0.dll
libgstplayer-1.0-0.dll
libgstreamer-1.0-0.dll
libgstriff-1.0-0.dll
libgstrtp-1.0-0.dll
libgstrtsp-1.0-0.dll
libgstsctp-1.0-0.dll
libgstsdp-1.0-0.dll
libgsttag-1.0-0.dll
libgsttranscoder-1.0-0.dll
libgsturidownloader-1.0-0.dll
libgstvideo-1.0-0.dll
libgstvulkan-1.0-0.dll
libgstwebrtc-1.0-0.dll
libgstwebrtcnice-1.0-0.dll
libgtk-4-1.dll
libharfbuzz-0.dll
libharfbuzz-cairo-0.dll
libharfbuzz-gobject-0.dll
libharfbuzz-icu-0.dll
libharfbuzz-subset-0.dll
libpango-1.0-0.dll
libpangocairo-1.0-0.dll
libpangoft2-1.0-0.dll
libpangowin32-1.0-0.dll
```

**Reply by badcel** [reply]
On a first glance `libgdk_pixbuf-2.0-0.DLL` is missing.

If this does not solve your problem I would kindly ask to check the dependencies yourself. You can do this via checking out the [windows gir files](https://github.com/gircore/gir-files/tree/main/windows). They define their dependencies in xml tags called "include" at the very top and there is the name of the expected DLL name in the "shared-library" xml attribute e.g. "shared-library="libgtk-4-1.dll" for the GTK-4.0.gir.

The dependencies for GTK would be:
```
<include name="Gdk" version="4.0"/>
  <include name="Gsk" version="4.0"/>
```

So you know to check the Gdk and Gsk gir files next.

Sometimes one binary can contain multiple other libraries like the GTK DLL containing gdk and gsk which is fine.

**Reply by PlatinumLucario** [reply]
Ah, just as I thought, there were some missing dependencies. lol

It wasn't just ``libgdk_pixbuff-2.0-0.dll`` that was missing, there was ``libgmodule-2.0-0.dll``, and there was one package that hadn't been installed: GtkSourceView5, which contains the much needed ``libgtksourceview-5-0.dll``.

Thank you so much!

**Comment by PlatinumLucario** [comment]
So after noticing that there was missing dependencies, despite saying that ``libgtk-4-1.dll`` was missing, I got the needed libraries into the directory, now it works. A big thank you to @badcel for pointing this out.

**Comment by KyodaiKen** [comment]
For me it works but I am basically doing it with a crowbar method because I currently just copy every DLL file in my MSYS2 environment's ucrt64 directory to the output directory. Is there a DLL list that has to be copied? I would like to only copy the DLL files needed for GirCore.Gio-2.0 and GirCore.Gtk-4.0.

Is there no easier way to create a list of DLLs to be copied?

**Reply by KyodaiKen** [reply]
I found a solution! You can use the following script to take care of building a DLL list to be copied. Run this script in the MSYS2 ucrt terminal. 

```bash
#!/bin/bash

# Configuration
BASE_URL="https://raw.githubusercontent.com/gircore/gir-files/main/windows"
UCRT_BIN="/ucrt64/bin"
START_GIRS=("Gtk-4.0" "Gio-2.0")

# State tracking
processed_girs=()
discovered_dlls=()

function is_processed() {
    local e match="$1"
    for e in "${processed_girs[@]}"; do [[ "$e" == "$match" ]] && return 0; done
    return 1
}

function extract_dependencies() {
    local gir_name="$1"
    if is_processed "$gir_name"; then return; fi
    processed_girs+=("$gir_name")

    echo "--> Analyzing GIR: $gir_name..." >&2

    local raw_content
    raw_content=$(curl -sL -f -r 0-65536 "$BASE_URL/$gir_name.gir")

    if [[ $? -ne 0 || -z "$raw_content" ]]; then
        echo "    [!] Could not fetch $gir_name.gir" >&2
        return
    fi

    # Extract Shared Libraries
    while read -r dll; do
        [[ -n "$dll" ]] && discovered_dlls+=("$dll")
    done < <(echo "$raw_content" | grep -oP 'shared-library="\K[^"]+' | tr ',' '\n')

    # Extract Includes
    while read -r sub_gir; do
        [[ -n "$sub_gir" ]] && extract_dependencies "$sub_gir"
    done < <(echo "$raw_content" | grep -oP '<include\s+name="[^"]+"\s+version="[^"]+"' | \
             sed -E 's/.*name="([^"]*)".*version="([^"]*)".*/\1-\2/')
}

# --- STAGE 1: GIR Discovery ---
echo "Stage 1: Discovering GObject-level DLLs..." >&2
for root in "${START_GIRS[@]}"; do
    extract_dependencies "$root"
done

# --- STAGE 2: Binary Deep Scan ---
echo -e "\nStage 2: Scanning for system dependencies via ntldd..." >&2
FINAL_LIST=()

# Deduplicate Stage 1 list first
unique_gir_dlls=($(printf "%s\n" "${discovered_dlls[@]}" | sort -u))

for dll in "${unique_gir_dlls[@]}"; do
    dll_path="$UCRT_BIN/$dll"
    if [[ -f "$dll_path" ]]; then
        # Add the main DLL
        FINAL_LIST+=("$dll")

        # Use sed to capture everything between "=> " and " ("
        # This is much more reliable than awk columns
        while read -r win_path; do
            if [[ -n "$win_path" ]]; then
                # Convert to unix path and get basename
                unix_path=$(cygpath -u "$win_path")
                # Only keep it if it's in our ucrt64 directory
                if [[ "$unix_path" == *"/ucrt64/"* ]]; then
                    FINAL_LIST+=("$(basename "$unix_path")")
                fi
            fi
        done < <(ntldd -R "$dll_path" | grep -i "ucrt64" | sed -n 's/.*=> \(.*\) (.*/\1/p')
    fi
done

# Final Cleanup: Deduplicate and format
DLL_LIST=$(printf "%s\n" "${FINAL_LIST[@]}" | sort -u | xargs)

echo "---" >&2
echo "Discovery complete." >&2
echo "DLL_LIST=\"$DLL_LIST\""
```
